doc/dev/adr/0020-error-modeling-headers.md

```diff
@@ -0,0 +1,103 @@
+
+In addition, users must be able to include arbitrary user properties in their command response to support this error reporting in case our standard fields are insufficient. This feature is likely in place already for all languages.
+
+On the command invoker side, we will add APIs for checking if a response was an application error and returning the error code and error data fields parsed from the MQTT message "AppErrCode" and "AppErrPayload" user properties.
+
+Similar to how our SDKs handle serializing the actual MQTT message payload, our SDKs will require the user provide the serializer for serializing/deserializing the AppErrPayload object.
+
+Other than these two new user properties, the over-the-wire behavior of our protocol won't change as a result of this decision.
+
+In order to provide a strongly-typed experience, we will also add codegen support for modeling both the possible error codes (enum-like list of possible values?) and the type of the error payload in DTDL. This modeling will be detailed in a separate ADR, though.
+
+By convention, the value of the AppErrPayload object should be a UTF-8 encoded JSON string since MQTT user property values must be UTF-8 encoded strings.
+
+## Code Example
+
+### Command executor side
+
+```csharp
+public ExtendedResponse<IncrementResponsePayload> Increment(IncrementRequestPayload request)
+{
+    if (request.IncrementValue < 0)
+    {
+        var response = new ExtendedResponse<IncrementResponsePayload>()
+        {
+            Response = new IncrementResponsePayload { CounterResponse = _counter },
+        };
+
+        // Specify error code, but no error payload  
+        response.WithApplicationError("negativeValue");
+
+        // Or you can specify error code and error payload
+        response.WithApplicationError(
+            "negativeValue",
+            new CustomErrorPayload() { ArgumentValue = request.IncrementValue },
+            new ErrorPayloadJsonSerializer(new TestEnvoys.Utf8JsonSerializer()));
+
+        return response;
+    }
+
+    // happy path omitted
+}
+```
+
+### Command invoker side
+
+```csharp
+public void main()
+{
+    MqttSessionClient mqttClient = ...;();
+    CounterClient counterClient = new CounterClient(mqttClient);
+
+    IncrementRequestPayload payload = new IncrementRequestPayload
+    {
+        IncrementValue = -1
+    };
+
+    var response = counterClient.Increment(executorId, payload).WithMetadata();
+    
+    // Check the RPC response for an application error
+    if (response.TryGetApplicationError(SomeJsonSerializer, out string? errorCode, out CustomErrorPayload? errorPayload))
```

Comment:
What is if the client use different Json(de-)Serializer?

Reply:
This proposal allows for user to have different serializers for handling error models in MQTT message payloads vs error models in MQTT user properties.